---
title: Summary of BGC analysis of other respiratory species
author: Rachel Myers, Ilya Zhbannikov 
date: "`r Sys.Date()`"
output: 
    html_document:
        toc: true
        self.contained: true
---

# Coryne genomes analysis
# Preliminaries

## Credits
Author: Rachel Myers, Ilya Zhbannikov

Investigator: Dr. Matthew Kelly

Date: `r Sys.Date()`

```{r recreate, echo = TRUE, eval = FALSE,  results = "asis"}

rm(list= ls(all= TRUE))
require("rmarkdown")
require("knitr")
library(stringr)

setwd("antismash/")

species_names1 <- list.dirs(path="final_genomes_2025-01-24/",full.names = F, recursive = F)
#species_names <- species_names1[which(species_names1 != "Moraxella_catarrhalis")]

species_names <- species_names1


render("Coryne_WGA_BGA_v4_summary.rmd", c("html_document"))

```



## Load packages and functions

Required r packages: 


* readxl
* tidverse
* ggplot2
* kableExtra
* table1

Set seed for reproducibility of random processes

```{r loadpackages, echo = F, warnings = F, error = F, message = F}
require(readxl)
require(ggplot2)
require(tidyverse)
require(kableExtra)
require(table1)
require(reshape2)
library(patchwork)
require(muscle)
require(ggmsa)
require(msa)
library(forcats)
library(openxlsx)
library(flextable)
set.seed(1234)
source("multiplot.R")

BGC_Orders = data.frame("Name" = c( 
  "NRPSandNRPSlike",
                "NRPS",
                "NRPS_like",
  "NAPAAs", 
                "NAPAA",
  "PKS",
                "T1PKS", 
                "T3PKS", 
                "transAT_PKS",
  "RiPPandRippLike",
                "lanthipeptide_class_i", 
                "lanthipeptide_class_ii",
                "lanthipeptide_class_iii",
                "lanthipeptide_class_iv",  
                "lanthipeptide_class_v", 
                "LAP", 
                "lassopeptide", 
                "linaridin", 
                "DUF692_associated_bacteriocin",
                "class_iid_bacteriocin",
                "RRE_containing",
                "thiopeptide",
                "RaS_RiPP",
                "RiPP_like",
                "sactipeptide",
                "epipeptide",
  "Terpenes", 
                "terpene",
  "Siderophores",
                "siderophore",
  "Other",
                "betalactone", 
                "ectoine", 
                "nucleoside", 
                "phenazine", "CDPS", "cyclic_lactone_autoinducer", "furan"),
                "Order" = seq(1,38),
                "level" = c("Family", "Child", "Child", # NRPSandNRPSlike
                            "Family", "Child", # NAPAAs
                            "Family",  "Child", "Child", "Child", # PKS
                    "Family", "Child", "Child", "Child","Child", "Child", "Child","Child", "Child","Child", "Child", "Child", "Child", "Child", "Child", "Child", "Child",# RiPP
                    "Family", "Child", # Terpenes
                    "Family", "Child", # Siderophores
                    "Family", "Child", "Child","Child", "Child", "Child", "Child", "Child"), #Other
                stringsAsFactors = FALSE
             )


BGC_map = list("NRPS"="NRPSandNRPSlike",
                "NRPS_like"="NRPSandNRPSlike",
                "NAPAA"="NAPAAs",
                "T1PKS"="PKS", 
                "T3PKS"="PKS", 
                "transAT_PKS"="PKS",
                "lanthipeptide_class_i"="RiPPandRippLike", 
                "lanthipeptide_class_ii"="RiPPandRippLike",
                "lanthipeptide_class_iii"="RiPPandRippLike",
                "lanthipeptide_class_iv"="RiPPandRippLike",  
                "lanthipeptide_class_v"="RiPPandRippLike", 
                "LAP"="RiPPandRippLike", 
                "lassopeptide"="RiPPandRippLike", 
                "linaridin"="RiPPandRippLike", 
                "DUF692_associated_bacteriocin"="RiPPandRippLike", 
                "class_iid_bacteriocin"="RiPPandRippLike", 
                "RRE_containing"="RiPPandRippLike", 
                "thiopeptide"="RiPPandRippLike", 
                "RaS_RiPP"="RiPPandRippLike", 
                "RiPP_like"="RiPPandRippLike", 
                "terpene"="Terpenes",
                "siderophore"="Siderophores",
                "betalactone"="Other", 
                "ectoine"="Other", 
                "nucleoside"="Other",
                "phenazine"="Other", 
                "CDPS"="Other", 
                "cyclic_lactone_autoinducer"="Other",
                "epipeptide"="RiPPandRippLike",
                "furan"="Other",
                "sactipeptide"="RiPPandRippLike")

```


```{r loaddata, echo = F, results = "asis", error = F, warning = F}

results <- list()
pmap <- read.xlsx("protocore_Family.xlsx", sheet = 1)
pmap2 <- read.xlsx("protocore_Family.xlsx", sheet = 3)

fh_metadata <- "FINAL_1.23.25_respiratory_genomes_AC.xlsx"
    

for(sn in species_names) {
    #print(sn)
  
    # meta data sheet - one row per strain / sample id
    #metadata <- read_excel(fh_metadata, sheet="Metadata", na="NA")
    metadata <- readxl::read_excel(fh_metadata, sheet=gsub("_", " ", sn), na="NA")
    ind <- which(metadata[,1]=="Reference")
    metadata <- metadata[-ind,]
    metadata <- metadata[which(!is.na(metadata[,1])), ]
    metadata <- metadata[order(metadata$strain), ]
    metadata$unknown_strains <- NULL
    if(sn=="Haemophilus_parainfluenzae") {
      metadata$biorepository_id <- metadata$`ans$accession`
    } else {
      metadata$biorepository_id <- metadata$ans_accession
    }
    #cat("<br>Metadata loaded")
    #print(metadata[which(duplicated(metadata$biorepository_id)==T), "biorepository_id"])
  
    if(sn == "Moraxella_catarrhalis") {
      results[[sn]] <- list(protocores=metadata, uniqueSeqs=NA, n=length(unique(metadata$biorepository_id)))
      next
    }
  
    basepath <- paste0("final_genomes_2025-01-24/", sn)
    fh_antismash <- paste0(basepath, "/antismash_out/All_antismashProto_cores.txt")
    
    
    #exclude some strains as requested
    #to_remove <- unique(read.csv("/Users/iz12/Library/CloudStorage/Box-Box/Coryne Genomics/Results/protocores_respiratory_to_remove.csv")$strain_id)
    #metadata <- metadata[ which(!(metadata$strain_id %in% to_remove)), ]

    protocores <- read.table(fh_antismash , sep = "\t", header= TRUE)
    #protocores$subjectid <- str_match(protocores$gbk_filename, "GCF_[0-9]*.[0-9]") # GCF_041450195.1
    protocores$subjectid <- basename(dirname(protocores$gbk_filename))
    
    #if(length(unique(protocores$subjectid)) != nrow(metadata)) {
    #  print("!!!")
    #  print(length(unique(protocores$subjectid)))
    #  print(nrow(metadata))
    #  print(sn)
    #  print(metadata$biorepository_id[which(!(metadata$biorepository_id %in% protocores$subjectid))])
    #  print(protocores$subjectid[which(!(protocores$subjectid %in% metadata$biorepository_id))])
    #  #stop()
    #}
    
    protocores <- merge(metadata, protocores, by.x = "biorepository_id", by.y = "subjectid")
    protocores$product_protoCore <- gsub("-", "_", protocores$product_protoCore )
    
    
    
    #protocores$product_protoCoreFamily <- pmap$product_protoCoreFamily[which(pmap$product_protoCore %in% protocores$product_protoCore)]
    protocores <- merge(protocores, pmap, by = "product_protoCore", all.x=T)
    #protocores <- merge(protocores, pmap2, by = "product_protoCore", all.x=T)
    
    ## manually update the RiPP_like product protoCore label to the granular version in gene_functions_CDS
    #sfilter <- which(protocores$product_protoCore == "RiPP_like" & grepl("DUF692", protocores$gene_functions_CDS))
    #protocores$product_protoCore[sfilter] <- "DUF692_associated_bacteriocin"
    #sfilter <- which(protocores$product_protoCore == "RiPP_like" & grepl("Lactococcin_972", protocores$gene_functions_CDS))
    #protocores$product_protoCore[sfilter] <- "class_iid_bacteriocin"
    
    # assign sequences unique identifiers 
    uniqueSeqs <- protocores[!duplicated(protocores$translation), c( "product_protoCore", "product_protoCoreFamily", "product_protoCoreFamilyColor", "translation",
                                                                     ifelse(sn=="Haemophilus_parainfluenzae", "organism$organism_name", "organism_name"), 
                                                                     "contig_edge", "strain","regionName", "gbk_filename"  )]
    
    uniqueSeqs$UniqueClusterID <- NA
    uniqueSeqs$SeqNumber <- 1
    
    for (ct in unique(uniqueSeqs$product_protoCore)) {
      rfilter <- which(uniqueSeqs$product_protoCore == ct)
      uniqueSeqs$SeqNumber[rfilter] <-cumsum(!duplicated(uniqueSeqs$translation[rfilter]))
      uniqueSeqs$UniqueClusterID[rfilter] <- gsub("-", "_", paste(ct, cumsum(!duplicated(uniqueSeqs$translation[rfilter])), sep = "_"))
    }
    
    protocores = merge(protocores, uniqueSeqs[, c( "product_protoCore", "translation", 
                                                   ifelse(sn=="Haemophilus_parainfluenzae", "organism$organism_name", "organism_name"), 
                                                   "contig_edge", "UniqueClusterID" , "SeqNumber")], all.x = TRUE)
    
    
    results[[sn]] <- list(protocores=protocores, uniqueSeqs=uniqueSeqs, n=length(unique(metadata$biorepository_id)))
    
    

}

results11 <- results
# Add pseudodiphtheriticum
#metadata <- readxl::read_excel("/Volumes/BioCAT/BioCAT_Projects/2022_KellyM_Coryne/Data/MetaCharacteristics/coryne_metadata_051823.xlsx")
#metadata <- metadata[which(metadata$final_species == "pseudodiphtheriticum"), ]
#to_remove <- unique(read.csv("/Users/iz12/Library/CloudStorage/Box-Box/Coryne Genomics/Results/protocores_respiratory_to_remove.csv")$strain_id)
#metadata <- metadata[ which(!(metadata$strain_id %in% to_remove)), ]


fh_antismash <- "protocores_respiratory_2025-01-27.csv"
protocores <- read.table(fh_antismash, sep = ",", header=T)
protocores <- protocores[which(protocores$final_species == "pseudodiphtheriticum"), c(-56,-57)] # removing pp family, family2 
protocores <- merge(protocores, pmap, by = "product_protoCore", all.x=T)
#protocores <- merge(protocores, pmap2, by = "product_protoCore", all.x=T)
# assign sequences unique identifiers 
uniqueSeqs <- protocores[!duplicated(protocores$translation), c( "product_protoCore", "product_protoCoreFamily", "product_protoCoreFamilyColor", "translation",
                                                                 "contig_edge", "strain_id","regionName", "gbk_filename"  )]
uniqueSeqs$UniqueClusterID <- NA
uniqueSeqs$SeqNumber <- 1
    
for (ct in unique(uniqueSeqs$product_protoCore)) {
      rfilter <- which(uniqueSeqs$product_protoCore == ct)
      uniqueSeqs$SeqNumber[rfilter] <-cumsum(!duplicated(uniqueSeqs$translation[rfilter]))
      uniqueSeqs$UniqueClusterID[rfilter] <- gsub("-", "_", paste(ct, cumsum(!duplicated(uniqueSeqs$translation[rfilter])), sep = "_"))
}
    
protocores = merge(protocores, uniqueSeqs[, c( "product_protoCore", "translation","contig_edge", "UniqueClusterID" , "SeqNumber")], all.x = TRUE)
    
results[["Corynebacterium_pseudodiphtheriticum"]] <- list(protocores=protocores, uniqueSeqs=uniqueSeqs, n=length(unique(protocores$biorepository_id)))
#results[["C. pseudodiphtheriticum"]] <- list(protocores=protocores, uniqueSeqs=uniqueSeqs, n=nrow(metadata))


## Add non-diphtheriae species BGCs
#nond <- c("Dolosigranulum_pigrum", "Haemophilus_influenzae", "Moraxella_catarrhalis", "Staphylococcus_aureus", "Staphylococcus_epidermidis", "Streptococcus_mitis",  "Streptococcus_pneumoniae")
#
#protocores <- results[[nond[1]]]$protocores[, c("product_protoCore", "product_protoCoreFamily", "product_protoCoreFamily2")]
#uniqueSeqs <- results[[nond[1]]]$uniqueSeqs
#n <- results[[nond[1]]]$n
#for(i in 2:length(nond)) {
#  sn <- nond[i]
#  protocores <- rbind(protocores, results[[sn]]$protocores[, c("product_protoCore", "product_protoCoreFamily", "product_protoCoreFamily2")])
#  uniqueSeqs <- rbind(uniqueSeqs, results[[sn]]$uniqueSeqs)
#  n <- n + results[[sn]]$n
#}
#
#results[["Non-diphtheriae species"]] <- list(protocores=protocores, uniqueSeqs=uniqueSeqs, n=n)
    


```

# Summary of BGC results


## Number of (any) BGCs per type, species

Overall, how many BGC clusters are identified ignoring BGC or unique protein sequence?

```{r bcgbytype_summaries, echo = FALSE, results = "asis", error = F, warning = F, message=F, eval=T}
dat <- results[[1]]$protocores[, c("product_protoCore", "product_protoCoreFamily", "product_protoCoreFamilyColor", "UniqueClusterID", "biorepository_id")]

splt <- strsplit(names(results)[1], "_")
dat$Species <- paste0(substr(splt[[1]][1], 1, 1), ". ", splt[[1]][2])
dat$Species <- paste0(dat$Species, "(", results[[1]]$n,")")

for(n in names(results)[2:(length(results)-1)]) {
  
  if(n == "Moraxella_catarrhalis") {
    tmp <- results[[n]]$protocores[,"biorepository_id", drop=F]
    tmp$product_protoCore <- tmp$product_protoCoreFamily <- tmp$product_protoCoreFamilyColor <- tmp$UniqueClusterID <- NA
    #tmp <- data.frame("product_protoCore"=NA, "product_protoCoreFamily"=NA, "product_protoCoreFamilyColor"=NA, "UniqueClusterID"=NA, "biorepository_id"="Moraxella_catarrhalis")
    tmp <- tmp[, c("product_protoCore", "product_protoCoreFamily", "product_protoCoreFamilyColor", "UniqueClusterID", "biorepository_id")]
  } else {
    tmp <- results[[n]]$protocores[, c("product_protoCore", "product_protoCoreFamily", "product_protoCoreFamilyColor", "UniqueClusterID", "biorepository_id")]
  }
  
  splt <- strsplit(n, "_")
  tmp$Species <- paste0(substr(splt[[1]][1], 1, 1), ". ", splt[[1]][2])
  tmp$Species <- paste0(tmp$Species, "(", results[[n]]$n,")")
  
   
  
  dat <- rbind(dat, tmp)
}



```

### BGC statistics

```{r, results='asis', echo=F, message=F, warning=F}

protocores_resp <- dat
write.csv(dat, file = "protocores_resp.csv")

protocores_resp <- protocores_resp[which(protocores_resp$product_protoCore != ""), ]
for(t in unique(protocores_resp$product_protoCore)) {
  protocores_resp$product_protoCoreFamily[which(protocores_resp$product_protoCore == t)] <- BGC_map[[t]]
}

ngs_by_class <- lapply(unique(protocores_resp$product_protoCore), function(x) {
    length(protocores_resp$"UniqueClusterID"[which(protocores_resp$product_protoCore == x)])
})
ngs_by_class2 <- lapply(unique(protocores_resp$product_protoCoreFamily), function(x) {
    length(protocores_resp$"UniqueClusterID"[which(protocores_resp$product_protoCoreFamily == x)])
})
names(ngs_by_class) = unique(protocores_resp$product_protoCore)
names(ngs_by_class2) = unique(protocores_resp$product_protoCoreFamily)

ngs_by_class = data.frame(rbind(
        do.call(rbind, ngs_by_class),
        do.call(rbind, ngs_by_class2)))
colnames(ngs_by_class) = "NoGeneClusters"
row.names(ngs_by_class)= gsub("_BGCCount", "", row.names(ngs_by_class))
ngs_by_class= merge(ngs_by_class, BGC_Orders, by.x = "row.names", by.y = "Name")
#ngs_by_class <- rbind(ngs_by_class, 
#                      c(Row.names="Total", NoGeneClusters=sum(ngs_by_class$NoGeneClusters[ngs_by_class$level=="Family"]), Order=max(ngs_by_class$Order)+1, level="Family"))
ngs_by_class$Row.names = factor(ngs_by_class$Row.names, levels = c(BGC_Orders$Name, "Total"))
row.names(ngs_by_class) = ngs_by_class$Row.names
ngs_by_class = ngs_by_class[order(ngs_by_class$Order), ]
indentset = which(ngs_by_class$level == "Child")
colnames(ngs_by_class) <- c("Type", "NoGeneClusters", "Order", "level")

z2 <- as.data.frame(t(ngs_by_class[ngs_by_class$level=="Child", c("Type", "NoGeneClusters"), drop = FALSE] %>% 
  summarise(across(where(is.numeric), c(~ sum(.x), 
                                        ~ median(.x), 
                                        ~ quantile(.x, probs = c(0.25)), 
                                        ~ quantile(.x, probs = c(0.75)), 
                                        ~ IQR(.x), 
                                        ~ mean(.x), 
                                        ~ sd(.x) ))))) #%>% 
#  mutate(Type = "Total")
rownames(z2) <- c("Total", "Median", "Q1", "Q3", "IQR", "Mean", "S.D.")
colnames(z2) <- "NoGeneClusters"
z2$Type <- rownames(z2)
z2 <- z2[,2:1]

bind_rows(ngs_by_class[, c("NoGeneClusters"), drop = FALSE], 
          z2[, "NoGeneClusters", drop=F] ) |> 
    #janitor::adorn_totals() %>%
    kbl(caption = "Number of all (not necessarily unique) gene clusters identified across all respiratory specimen", row.names = T) %>%
    kable_classic(full_width = F,  position = "left", html_font = "Cambria") %>%
    add_indent(indentset) 


bind_rows(ngs_by_class[, c("Type", "NoGeneClusters"), drop = FALSE], 
          z2) |> 
  flextable::flextable() |>
    align(part = "all") |> # left align
    set_caption(caption = "Table S4. Number of all (not necessarily unique) gene clusters identified across all respiratory specimen") |>
    font(fontname = "Times New Roman", part = "all") |> 
    fontsize(size = 10, part = "body") |> 
    theme_booktabs() |> # default theme
    autofit() |> 
    hline(i = ~ before(Type, Type), border = fp_border_default()) |>
    padding(i=indentset, j=1, padding.left=20) |>
    save_as_docx(path="S4_OtherRespStrains_BGC.docx")




```

```{r, echo = FALSE, results = "asis", error = F, warning = F, message=F}
protocores_resp <- dat

temp <- data.frame(table(protocores_resp$biorepository_id))
df_mrx <- protocores_resp[which(protocores_resp$Species == "M. catarrhalis(41)"), "biorepository_id", drop=F]
df_mrx$Freq <- 0
colnames(df_mrx) <- c("Var1", "Freq")
temp <- rbind(temp, df_mrx)
cat("<br>Mean and median (interquartile range) of BGCs per specimen<br>")
ans <- quantile(temp$Freq, type = 5)
cat("Mean:", mean(temp$Freq), ", S.D.:", sd(temp$Freq), ", Median:", ans[3],  "(", ans[2], ", ", ans[4], ")\n")


```

```{r, echo = FALSE, results = "asis", error = F, warning = F, message=F, fig.width=5, fig.height=5}
#dat_sum <- as.data.frame(dat %>% dplyr::group_by(BGCType=product_protoCore)  %>% dplyr::summarise(count=n()) %>% arrange(-count))
dat_sum <- as.data.frame(dat %>% dplyr::group_by(BGCType=product_protoCoreFamily)  %>% dplyr::summarise(count=n()) %>% arrange(-count))
dat_sum <- dat_sum[complete.cases(dat_sum), ]
p <- ggcharts::bar_chart(dat_sum, BGCType, count) + theme_bw() + xlab("BGC type") + ylab("Number of non-unique BGCs") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
p
```

## By species analyzed



```{r, echo = FALSE, message=FALSE, warning=F, results = "asis", error = F, warning = F}
# * Non-diphtheriae species: `r nond`
#dat_sum <- as.data.frame(dat %>% dplyr::group_by(BGCType=product_protoCore, Species=Species)  %>% dplyr::summarise(count=n()) %>% arrange(-count))
dat_sum <- as.data.frame(dat %>% dplyr::group_by(BGCType=product_protoCoreFamily, Species=Species)  %>% dplyr::summarise(count=n()) %>% arrange(-count))
dat_sum <- dat_sum[complete.cases(dat_sum), ]
#dat_sum <- dat_sum[order(dat_sum$count, decreasing = T),]
dat_sum$BGCType <- factor(dat_sum$BGCType, levels=levels(p$data$BGCType))
# Barplot
p <- ggplot(dat_sum, aes(x=BGCType, y=count, fill = Species)) + geom_bar(stat="identity") + theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
p <- p + coord_flip() + theme_bw() + xlab("BGC type") + ylab("Number of non-unique BGCs") #+ scale_fill_manual(na.translate = F)
p

dat_sum1 <- dat_sum
```
```{r, echo = FALSE, message=FALSE, warning=F, results = "asis", error = F, fig.width=5, fig.height=5}
dat <- results[[1]]$protocores[, c("product_protoCore", "product_protoCoreFamily", "product_protoCoreFamilyColor")]

splt <- strsplit(names(results)[1], "_")
dat$Species <- paste0(substr(splt[[1]][1], 1, 1), ". ", splt[[1]][2])
dat$Species <- paste0(dat$Species, "(", results[[1]]$n,")")

for(n in names(results)[2:length(results)]) {
  
  if(n == "Moraxella_catarrhalis") {
    tmp <- data.frame("product_protoCore"="", "product_protoCoreFamily"="",  "product_protoCoreFamilyColor"="white")
  } else {
    tmp <- results[[n]]$protocores[, c("product_protoCore", "product_protoCoreFamily", "product_protoCoreFamilyColor")]
  }
  
  if(n != "Non-diphtheriae species") {
    splt <- strsplit(n, "_")
    tmp$Species <- paste0(substr(splt[[1]][1], 1, 1), ". ", splt[[1]][2])
  } else {
    tmp$Species <- n
  }
  
  tmp$Species <- paste0(tmp$Species, "(", results[[n]]$n,")")
  
  
  
  dat <- rbind(dat, tmp)
}

dat_sum <- as.data.frame(dat %>% dplyr::group_by(Species=Species)  %>% dplyr::summarise(count=n()) %>% arrange(-count))
# Barplot
p2 <- ggcharts::bar_chart(dat_sum, Species, count) + theme_bw() + xlab("BGC type") + ylab("Number of non-unique BGCs") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
p2


#[which(dat$product_protoCore %in% types_to_include),]
```


* By species and BGC type

```{r, echo = FALSE, message=FALSE, warning=F, results = "asis", error = F, fig.width=7, fig.height=5}
types_to_include <- c("", as.character(unique(dat_sum1$BGCType[which(dat_sum1$count >= 0)])))


dat_sum <- as.data.frame(dat[which(dat$product_protoCoreFamily %in% types_to_include), ] %>% dplyr::group_by(Species=Species, BGCType=product_protoCoreFamily, Color=product_protoCoreFamilyColor)  %>% dplyr::summarise(count=n()))
dat_sum$Species <- factor(dat_sum$Species, levels=levels(p2$data$Species))
p <- ggplot(dat_sum, aes(x=Species, y=count, fill = BGCType)) + geom_bar(stat="identity")
p <- p + coord_flip() + theme_bw() + scale_fill_manual(breaks = dat_sum$BGCType , values=dat_sum$Color, na.translate = F) + theme_bw() + xlab("BGC type") + ylab("Number of non-unique BGCs") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
p <- p + geom_col(position = position_stack(reverse = TRUE))
p

```


* Mean number of BGCs

```{r, echo = FALSE, message=FALSE, warning=F, results = "asis", error = F, fig.width=7, fig.height=5}
dat <- results[[1]]$protocores[, c("product_protoCore", "product_protoCoreFamily", "product_protoCoreFamilyColor")]

splt <- strsplit(names(results)[1], "_")
dat$Species <- paste0(substr(splt[[1]][1], 1, 1), ". ", splt[[1]][2])
dat$Species <- paste0(dat$Species, "(", results[[1]]$n,")")
dat$n <- results[[1]]$n

for(n in names(results)[2:length(results)]) {
  
  if(n == "Moraxella_catarrhalis") {
    tmp <- data.frame("product_protoCore"="", "product_protoCoreFamily"="", "product_protoCoreFamilyColor"="white")
  } else {
    tmp <- results[[n]]$protocores[, c("product_protoCore", "product_protoCoreFamily", "product_protoCoreFamilyColor")]
  }
  
  if(n != "Non-diphtheriae species") {
    splt <- strsplit(n, "_")
    tmp$Species <- paste0(substr(splt[[1]][1], 1, 1), ". ", splt[[1]][2])
  } else {
    tmp$Species <- n
  }
  
  tmp$Species <- paste0(tmp$Species, "(", results[[n]]$n,")")
  tmp$n <- results[[n]]$n
  
  
  
  dat <- rbind(dat, tmp)
}


dat_sum <- as.data.frame(dat %>% 
                           dplyr::group_by(Species=Species, BGCType=product_protoCoreFamily, Color=product_protoCoreFamilyColor, n=n)  %>%
                           dplyr::summarise(count=n())) 

dat_sum$mn <- dat_sum$count/dat_sum$n

  
  
#dat_sum <- as.data.frame(dat_sum %>% 
#                           dplyr::group_by(Species, BGCType, Color)  %>%
#                           dplyr::summarise(mn=mean(count))) 


dat_sum2 <- dat_sum %>% dplyr::group_by(Species) %>% summarise(sum=sum(mn))

#dat_sum$Species <- factor(dat_sum$Species, levels=levels(p2$data$Species))
dat_sum$Species <- factor(dat_sum$Species, levels=dat_sum2[order(dat_sum2$sum, decreasing = F), ]$Species)


p <- ggplot(dat_sum, aes(x=Species, y=mn, fill = BGCType)) + geom_bar(stat="identity")
p <- p + coord_flip() + theme_bw() + scale_fill_manual(breaks = dat_sum$BGCType , values=dat_sum$Color, na.translate = F) + theme_bw() + xlab("BGC type") + ylab("Mean number of BGCs") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
p <- p + geom_col(position = position_stack(reverse = TRUE))
p

```



## Bacterial genome statistics

```{r, results='asis', echo=F, message=F, warning=F, eval=T}

results <- list()

for(sn in species_names1) {
    basepath <- paste0("final_genomes_2025-01-24/", sn)
    
    metadata <- readxl::read_excel(fh_metadata, sheet=gsub("_", " ", sn), na="NA")
    ind <- which(metadata[,1]=="Reference")
    metadata <- metadata[-ind,]
    metadata <- metadata[which(!is.na(metadata[,1])), ]
    metadata <- metadata[order(metadata$strain), ]
    metadata$unknown_strains <- NULL
    if(sn=="Haemophilus_parainfluenzae") {
      metadata$biorepository_id <- metadata$`ans$accession`
    } else {
      metadata$biorepository_id <- metadata$ans_accession
    }
    #
    
    fastafh <- list.files(paste0(basepath, "/fasta/"), pattern = "fna", full.names = TRUE )
    included <- sapply(1:length(metadata$biorepository_id), function(i) grep(metadata$biorepository_id[i], fastafh))
    fastafh <- fastafh[included]
    

    fastadata1 = lapply(fastafh, seqinr::read.fasta )
    ln <- sapply(seqinr::getLength(fastadata1), sum)

    #cat("<br>Mean genome length: ", mean(ln), ", s.d.: ", sd(ln), ", Min, max:", min(ln), ", ", max(ln), "\nTotal genomes: ", length(ln), "</br>")
    
    results[[gsub("_", " ", sn)]] <- list(Mean=mean(ln), SD=sd(ln), Min=min(ln), Max=max(ln), Total_genomes=length(ln))
    
    #temp <- data.frame(table(protocores$biorepository_id))
    #cat("<br>Mean and median (interquartile range) of BGCs per specimen<br>")
    #ans <- quantile(temp$Freq, type = 5)
    #cat("Mean:", mean(temp$Freq), ", Median:", ans[3],  "(", ans[2], ", ", ans[4], ")\n")

}


results <- do.call(rbind, results)
knitr::kable(results)

```


